### DISCUSSION TITLE
onClick on overlapping areas triggers all, even with different z

### DISCUSSION THREAD

**Discussion by pjoe** [replit/kaboom#721] [Q&A]
Using kaboom 3000.0.3.

Expect: `onClick` only to trigger on topmost GameObj if there are overlapping areas, especially if they have different z

Actual: All `onClick` handlers matching mousepos are triggered

Example: https://replit.com/@pjoePanda/kaboom3000-area-click-overlap

**Comment by pjoe** [comment]
FWIW: doing this as work around:

```ts
  onClick(() => {
    const pos = mousePos();
    const match = get("tile").reduceRight<GameObj | undefined>((acc, go) => {
      if (acc && acc.z >= go.z) return acc;
      if (go.hasPoint(pos)) return go;
      return acc;
    }, undefined);
    if (match) {
      console.debug(`match id:${match.id}, type:${match.tileType}`);
      match.destroy();
    }
  });
```

**Comment by slmjkdbtl** [comment]
This is the expected behavior, you can check [this example](https://github.com/replit/kaboom/blob/b71dd8aa3986ea110d97bfa02a41da050c6b2a0d/examples/drag.js#L48-L60) which iterates objects in reverse z order to check for mouse click

**Reply by pjoe** [reply]
Thanks, will `get(tag)` always return the list in z order?

**Reply by slmjkdbtl** [reply]
no, the list is the order of object being added, you can sort manually like
```js
get(tag).sort((o1, o2) => (o1.z ?? 0) - (o2.z ?? 0))
```

**Reply by pjoe** [reply]
Can then simplify to:
```ts
  onClick(() => {
    const clicked = get("tile").findLast((go) => go.isHovering());
    if (clicked) {
      clicked.destroy();
    }
  });
```

**Reply by pjoe** [reply]
Looks like z sorting already happens in update: https://github.com/replit/kaboom/blob/master/src/kaboom.ts#L3026

**Reply by slmjkdbtl** [reply]
Yep z sorting is done when updating / drawing objects, but not in `get()`

**Reply by pjoe** [reply]
Ok, however the sort inside update is "inplace" as far as I can see, so it persists after the the update. At least for my use case this is good enough as I'm not rapidly adding new objects and calling get between updates.

**Reply by slmjkdbtl** [reply]
Oh yes you are right, it does persist every frame
